### PR TITLE
Week6

### PR DESCRIPTION
**Pull request by JolandaNava**

### CODE REVIEW

File: src/Week6/Fibonacci.hs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+streamToList :: Stream a -> [a]
+streamToList (Stream x sx) = x : streamToList sx
+
+instance Show a => Show (Stream a) where
+    show = show . take 40 . streamToList
+
+---------- ex 4 ----------
+streamRepeat :: a -> Stream a
+streamRepeat x = Stream x (streamRepeat x)
```
Comment (chshersh) [inline src/Week6/Fibonacci.hs]: I mentioned previosly on one of my lectures about lazy evaluation that the following implementation for function like repeat is inefficient.

```haskell
repeat x = x : repeat x
```

Instead, it's more efficient to write:

```haskell
repeat x = let xs = x : xs in xs
```

You could to implement `streamRepeat` in similar way 🙂

File: src/Week6/Fibonacci.hs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            c21 = a21*b11 + a22*b21
+            c22 = a21*b12 + a22*b22
+
+f :: Matrix
+f = ((1,1),(1,0))
+
+fib4 :: Integer -> Integer
+-- fib4 0 = 0
+-- I can't specify the 0th case if I want to eta reduce fib4
+fib4 = snd . fst . (^) f
```
Comment (chshersh) [inline src/Week6/Fibonacci.hs]: Yeah, it's a Haskell restriction. You either eta-reduce each case or none of them.

**Review by chshersh — APPROVED** [review]
This looks pretty good!

File: src/Week6/Fibonacci.hs [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    where
+        streamOfStreams :: Stream (Stream Integer)
+        streamOfStreams = streamMap streamRepeat nats
+
+        recursivelyInterleave :: Stream (Stream Integer) -> Stream Integer
+        recursivelyInterleave (Stream s ss) = interleaveStreams s $ recursivelyInterleave ss
+
+---------- ex 6 ----------
+x :: Stream Integer
+x = Stream 0 $ Stream 1 $ streamRepeat 0
```
Comment (arbus) [inline src/Week6/Fibonacci.hs:59]: This is still the most mind blowing line of Haskell for me

**Review by arbus — APPROVED** [review]

**Pull request merged**